### PR TITLE
(PDB-2230) support submission of certname/command/version via post params

### PR DESCRIPTION
**Pull request by wkalt**
This allows us to do more granular logging at the wire without parsing the
command payload itself.

### CODE REVIEW

File: documentation/api/command/v1/commands.markdown [hunk 1]
```diff
@@ -166,6 +180,14 @@ To post a `replace facts` command you can use the following curl command:
 An example of `deactivate node`:
 
     curl -X POST \
+      -H 'Content-Type:application/json' \
+      -H 'Accept:application/json' \
+      -d '{"certname":"test1","producer_timestamp":"2015-01-01"}' \
```
Comment (ajroetker) [inline documentation/api/command/v1/commands.markdown]: Can we not make `certname` in the command body optional?

File: puppet/spec/unit/indirector/catalog/puppetdb_spec.rb [hunk 2]
```diff
@@ -39,11 +39,7 @@ def save
 
     it "should POST the catalog command as a JSON string" do
       command_payload = subject.munge_catalog(catalog, options)
-      payload = {
-        :command => Puppet::Util::Puppetdb::CommandNames::CommandReplaceCatalog,
-        :version => 7,
-        :payload => command_payload,
-      }.to_json
+      payload = command_payload.to_json
```
Comment (ajroetker) [inline puppet/spec/unit/indirector/catalog/puppetdb_spec.rb]: what the need for the extra step? can we just `to_json` the line above?

File: puppet/spec/unit/indirector/facts/puppetdb_spec.rb [hunk 3]
```diff
@@ -50,11 +50,7 @@ def save
         "producer_timestamp" => "a test",
       }
 
-      payload = {
-        :command => CommandReplaceFacts,
-        :version => 4,
-        :payload => f,
-      }.to_pson
+      payload = f.to_pson
```
Comment (ajroetker) [inline puppet/spec/unit/indirector/facts/puppetdb_spec.rb]: Similarly here, do we need `f`?

File: src/puppetlabs/puppetdb/cheshire.clj [hunk 4]
```diff
@@ -19,23 +19,29 @@
...
+  (generate/add-encoder
+    org.joda.time.DateTime
+    (fn [data ^JsonGenerator jsonGenerator]
+      (.writeString jsonGenerator (to-string data))))
   (generate/add-encoder
-   org.joda.time.DateTime
-   (fn [data ^JsonGenerator jsonGenerator]
-     (.writeString jsonGenerator (to-string data)))))
+    RawJsonString
+    (fn [data ^JsonGenerator jsonGenerator]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/cheshire.clj]: @senior do we have to worry about reflection with defrecord?

File: src/puppetlabs/puppetdb/http/command.clj [hunk 5]
```diff
@@ -21,21 +21,31 @@
 
 (def valid-commands-str (str/join ", " (sort (vals command-names))))
 
+(defn coerce-to-int
+  [s]
+  (if (number? s) s (Integer. s)))
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: Can you use https://github.com/puppetlabs/clj-kitchensink/blob/master/src/puppetlabs/kitchensink/core.clj#L207 here or parse-number?

File: src/puppetlabs/puppetdb/http/command.clj [hunk 6]
```diff
@@ -21,21 +21,31 @@
...
+
 (defn validate-command-version
   [app]
-  (fn [{:keys [body-string] :as req}]
-    (let [{:keys [command version]} (json/parse-string body-string true)
+  (fn [{:keys [body-string params param-post?] :as req}]
+    (let [{:strs [command version]} (if param-post?
+                                      params
+                                      (json/parse-string body-string))
+          command (str/replace command "-" " ")
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: So it seems kind of silly to go back and forth between `-` and `" "`, would we have to bump versions of anything to just always use `-` in our commands?

Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: I.e. would this be a good patch to do that in?

File: src/puppetlabs/puppetdb/utils.clj [hunk 7]
```diff
@@ -13,7 +13,8 @@
             [slingshot.slingshot :refer [try+ throw+]]
             [com.rpl.specter :as sp])
   (:import [java.net MalformedURLException URISyntaxException URL]
-           [org.postgresql.util PGobject]))
+           [org.postgresql.util PGobject]
+           [puppetlabs.puppetdb.cheshire RawJsonString]))
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/utils.clj]: I think the spaces in these imports are unnecessary, `[puppetlabs.puppetdb.cheshire RawJsonString]` means import `puppetlabs.puppetdb.cheshire` and `puppetlabs.puppetdb.cheshire.RawJsonString`, I think you just want `[puppetlabs.puppetdb.cheshire.RawJsonString]`.

File: src/puppetlabs/puppetdb/middleware.clj [hunk 8]
```diff
@@ -293,10 +293,19 @@
...
+              reconstructed-body (assoc params "payload" body-string)
+              param-post? (boolean (and (get params "certname")
+                                        (get params "version")
+                                        (get params "command")))]
           (if (nil? body-string)
             (http/error-response (str "Empty body for application/json submission"))
-            (app (assoc req :body-string body-string))))
+            (let [submission-body (if param-post?
+                                    (-> params
+                                        (assoc "payload" body-string)
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/middleware.clj]: looks like you could just use `reconstructed-body` here

File: src/puppetlabs/puppetdb/middleware.clj [hunk 9]
```diff
@@ -293,10 +293,19 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))
         "application/json"
-        (let [body-string (request/body-string req)]
+        (let [body-string (request/body-string req)
+              reconstructed-body (assoc params "payload" body-string)
+              param-post? (boolean (and (get params "certname")
+                                        (get params "version")
+                                        (get params "command")))]
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/middleware.clj]: It feels like if they specify any one of these, that we should error if they don't have all of them.

File: src/puppetlabs/puppetdb/utils.clj [hunk 10]
```diff
@@ -310,6 +311,12 @@
                 #(update % 0 underscores->dashes)
                 m))
 
+(defn synthesize-body-str
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/utils.clj]: Could we call this `cmd-params->json-str` or something more specific?

File: src/puppetlabs/puppetdb/utils.clj [hunk 11]
```diff
@@ -310,6 +311,12 @@
...
+  [{:strs [command version certname payload]}]
+  (json/generate-string
+    {:command (string/replace command "-" " ") :version (Integer. version)
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/utils.clj]: It looks like we're doing this twice, in `validate-command...` and here, is it possible to consolidate?

File: documentation/api/command/v1/commands.markdown [hunk 12]
```diff
@@ -32,19 +32,24 @@ The entire command MUST be encoded as UTF-8.
...
+two mechanisms:
 
-* A `POST` is used
-* The `POST` body must contain the JSON payload.
-* There is an `Accept` header that matches `application/json`.
-* The content-type is `application/json`.
+* Payload and parameters: This method entails POSTing the certname, command name,
+command version, and optionally the checksum as parameters, with the POST body
+containing the given command's wire format. This mechanism allows PuppetDB to
+provide some information at the wire without inspecting the command payload
```
Comment (ajroetker) [inline documentation/api/command/v1/commands.markdown]: This mechanism allows PuppetDB to provide better validation and feedback at the time of POSTing without...

File: src/puppetlabs/puppetdb/cheshire.clj (hunk 4 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/cheshire.clj]: @ajroetker The question should be, do we have to worry about it with 'extend'. My gut tells me no, since the compiler should statically know that the type of the `this` param (data in this case) is RawJsonString. But the Clojure compiler often doesn't understand things that we'd consider obvious. The best way to find out is to set `*warn-on-reflection*` and see what happens.

File: documentation/api/command/v1/commands.markdown (hunk 1 above)
Comment (wkalt) [inline documentation/api/command/v1/commands.markdown]: I think we should avoid this for now. We'll end up having to write that it's optional in one case but not the other, which starts to feel like a separate version. I'd vote for leaving it until we can bump the command and retire the old mechanism.

File: puppet/spec/unit/indirector/catalog/puppetdb_spec.rb (hunk 2 above)
Comment (wkalt) [inline puppet/spec/unit/indirector/catalog/puppetdb_spec.rb]: will fix

File: puppet/spec/unit/indirector/facts/puppetdb_spec.rb (hunk 3 above)
Comment (wkalt) [inline puppet/spec/unit/indirector/facts/puppetdb_spec.rb]: nope

File: src/puppetlabs/puppetdb/cheshire.clj (hunk 4 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/cheshire.clj]: I'll add a type hint

File: src/puppetlabs/puppetdb/http/command.clj (hunk 5 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/command.clj]: sure, I don't mind

File: src/puppetlabs/puppetdb/http/command.clj (hunk 6 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/command.clj]: I'd be on board with this change, but I feel like it's enough of a docs/code change for its own patch. It would break cross-version imports from 3->4 (which we don't guarantee) without some kind of replace call like this one -- don't know that we care.

File: src/puppetlabs/puppetdb/utils.clj (hunk 7 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/utils.clj]: this doesn't seem to work

File: test/puppetlabs/puppetdb/middleware_test.clj [hunk 13]
```diff
@@ -117,13 +118,26 @@
...
     (testing "url encoded payload should populate body-string"
       (let [test-req {:params {"payload" test-content}
                       :headers {"content-type" "application/x-www-form-urlencoded"}}]
         (is (= (wrapped-fn test-req)
-               (assoc test-req :body-string test-content)))))))
+               (assoc test-req :body-string test-content)))))
+
+    (testing "param-post? is true when params are included"
+      (let [test-req {:params {"certname" "foo.com"
+                               "command" "fix drywall"
```
Comment (mullr) [inline test/puppetlabs/puppetdb/middleware_test.clj]: :+1:

File: src/puppetlabs/puppetdb/utils.clj (hunk 7 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/utils.clj]: sorry, leave out the `[]` it should just be `puppetlabs.puppetdb.cheshire.RawJsonString`, https://clojuredocs.org/clojure.core/import

File: src/puppetlabs/puppetdb/http/command.clj (hunk 6 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: :+1: we can make a ticket and put it in another patch

Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: Ticket made PDB-2311, but a note from the channel... Should these be underscores instead of dashes?

**Comment by ajroetker** [comment]
@puppetlabs-jenkins test this please

File: src/puppetlabs/puppetdb/utils.clj [hunk 14]
```diff
@@ -310,6 +311,12 @@
                 #(update % 0 underscores->dashes)
                 m))
 
+(defn cmd-params->json-str
+  [{:strs [command version certname payload]}]
+  (json/generate-string
+    {:command (string/replace command "-" " ") :version (Integer. version)
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/utils.clj]: looks like this `string/replace` is unnecessary now

File: src/puppetlabs/puppetdb/middleware.clj [hunk 15]
```diff
@@ -286,10 +286,30 @@
...
+            (nil? body-string)
             (http/error-response (str "Empty body for application/json submission"))
-            (app (assoc req :body-string body-string))))
+
+            :else
+            (if param-post?
+              (let [command' (s/replace command "_" " ")
+                    submission-body (-> params
+                                        (assoc "payload" body-string)
+                                        (update "command" s/replace "_" " ")
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/middleware.clj]: can this just be `assoc "command" command'`?

File: src/puppetlabs/puppetdb/http/command.clj [hunk 16]
```diff
@@ -21,21 +21,30 @@
 
 (def valid-commands-str (str/join ", " (sort (vals command-names))))
 
+(defn coerce-to-int
+  [s]
+  (if (number? s) s (Integer/parseInt s)))
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: looks like my other comment got deleted, but I think `puppetlabs.kitchensink.core/parse-int` does basically the same thing as what we want here

Comment (wkalt) [inline src/puppetlabs/puppetdb/http/command.clj]: no, it requires a string as input

File: src/puppetlabs/puppetdb/middleware.clj [hunk 17]
```diff
@@ -286,10 +286,30 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))
         "application/json"
```
Comment (ajroetker) [inline src/puppetlabs/puppetdb/middleware.clj]: It seems like this function and https://github.com/puppetlabs/puppetdb/pull/1807/files#diff-d82b3486b4cd22472cd542b7773086e9L24 could/should probably be combined into 1 function

File: src/puppetlabs/puppetdb/http/command.clj (hunk 16 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/http/command.clj]: It looks like below you're using `Integer.` for this, could you the same thing be done in both places?

File: src/puppetlabs/puppetdb/middleware.clj [hunk 18]
```diff
@@ -286,10 +286,30 @@
...
-        (let [body-string (request/body-string req)]
-          (if (nil? body-string)
+        (let [{:strs [certname version command]} params
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/middleware.clj]: If I'm not misunderstanding, we already have a command-app handler that ensures the incoming request will be application/jjson, so I think we might be able to considerably simplify this function and the related test function.

File: src/puppetlabs/puppetdb/middleware.clj [hunk 19]
```diff
@@ -286,10 +286,30 @@
...
+              body-string (request/body-string req)
+              param-post? (boolean (and certname version command))]
+
+          (cond
+            (and (not param-post?) (or certname version command))
+            (http/error-response (str "Command submission via POST parameters requires certname, version, and command."))
```
Comment (rbrw) [inline src/puppetlabs/puppetdb/middleware.clj]: Possibly vestigial "str" here and in the other error-response.

Comment (rbrw) [inline src/puppetlabs/puppetdb/middleware.clj]: Also wondered about a possibly more direct message like "Commands submitted via POST require certname, version, and command parameters." or similar.

File: src/puppetlabs/puppetdb/http/command.clj (hunk 5 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/http/command.clj]: just to tie this off, neither of those works here. I also let payload-to-body-string be with respect to your suggested refactor because rob has some code on top of this that removes the function anyway.

**Comment by wkalt** [comment]
@ajroetker comments addressed

**Comment by ajroetker** [comment]
I'm :+1: everything lgtm and works

File: src/puppetlabs/puppetdb/middleware.clj [hunk 20]
```diff
@@ -284,15 +284,36 @@
...
+              (let [command' (s/replace command "_" " ")
+                    submission-body (-> params
+                                        (assoc "payload" body-string)
+                                        (assoc "command" command')
+                                        utils/cmd-params->json-str)]
+                (-> req
+                    (assoc-in [:params "command"] command')
+                    (assoc :body-string submission-body :param-post? param-post?)
+                    app))
+              (app (assoc req :body-string body-string :param-post? param-post?)))))
```
Comment (senior) [inline src/puppetlabs/puppetdb/middleware.clj]: This seems like a good candidate for being a handler, not middleware, it should be addressed as a separate issue though.

File: src/puppetlabs/puppetdb/utils.clj [hunk 21]
```diff
@@ -13,7 +13,8 @@
             [slingshot.slingshot :refer [try+ throw+]]
             [com.rpl.specter :as sp])
   (:import [java.net MalformedURLException URISyntaxException URL]
-           [org.postgresql.util PGobject]))
+           [org.postgresql.util PGobject]
+           puppetlabs.puppetdb.cheshire.RawJsonString))
```
Comment (senior) [inline src/puppetlabs/puppetdb/utils.clj]: cosmetic, by why have this import line look different from the others?

**Comment by senior** [comment]
:+1:

File: src/puppetlabs/puppetdb/utils.clj (hunk 21 above)
Comment (ajroetker) [inline src/puppetlabs/puppetdb/utils.clj]: @senior `[org.postgresql.util PGobject]` will import `org.postgresql.util` and `PGobject`, here we just want `RawJsonString`

Comment (senior) [inline src/puppetlabs/puppetdb/utils.clj]: Now that I look closer, this isn't needed at all. That line can be removed and line 318 can be changed to (json/->RawJsonString payload)
